text/0000-impl-trait-expressions.md

```diff
@@ -0,0 +1,148 @@
+- Feature Name: impl-trait-expressions
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Rust closures allow the programmer to create values of anonymous types which
+implement the `Fn*` traits. This RFC proposes a generalisation of this feature
+to other traits. The syntax looks like this:
+
+```rust
+fn main() {
+    let world = "world";
+    let says_hello_world = impl fmt::Display {
+        fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+            write!(f, "hello {}", world)
+        }
+    };
+
+    println!("{}", says_hello_world);
+}
+```
+
+# Motivation
+[motivation]: #motivation
+
+Sometimes we need to create a once-off value which implements some trait,
+though having to explicitly declare a type in these situations can be
+unnecessarily painful and noisy. Closures are a good example of how
+this problem can be ameliorated by adding the ability to declare once-off
+values of anonymous types.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+`impl Trait` expressions allow the user to create a value of an anonymous type which
+implements `Trait`. These expressions behave the same as closures - they
+capture local variables, by either move or reference, and collect them into an
+anonymous struct which implements the required trait.
+
+To better understand the behaviour of `impl Trait`, consider the following code:
+
+```rust
+let y = String::from("hello");
+let foo = move || println!("{}", y);
+```
+
+With this RFC, the above code becomes syntax sugar for:
+
+```rust
+let y = String::from("hello");
+let foo = move impl FnOnce<()> {
+    type Output = ();
+
+    extern "rust-call" fn call_once(self, args: ()) {
+        println!("{}", y);
+    }
+};
+```
+
+Which, in turn, is syntax sugar for:
+
+```rust
+let y = String::from("hello");
+
+struct MyCoolAnonType {
+    y: String,
+}
+
+impl FnOnce<()> for MyCoolAnonType {
+    type Output = ();
+
+    extern "rust-call" fn call_once(self, args: ()) {
+        println!("{}", self.y);
+    }
+}
+
+let foo = MyCoolAnonType { y };
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This feature is fully described in the guide-level explanation. As this is a
+generalisation of the existing closure syntax I suspect that the implementation
+would be fairly straight-forward.
```

Comment:
I'd like at least the following to be discussed:

1. What is the interaction with https://github.com/rust-lang/rfcs/pull/2229?

2. What changes are there to the grammar? and are there any ambiguities (there are, see below)

3. What happens when I implement the `Fn` trait and it is a subtrait once removed of `FnOnce`? Do I also get the supertrait impls?

4. How do I implement several traits at once? Can I do that? This ties into 3.

5. What is the type of an anonymous impl? A voldemort type?

6. Will `Copy` and `Clone` be implemented for the impl trait expression if possible like for closures?

7. What happens if I write `let obj = impl Foo<'_> { ... };`? Normally this would quantify a lifetime `'a` so we'd have `impl<'a> Foo<'a> for X { ... }`...

8. What rules if any apply for turning `let x = impl Foo { ... }` into a trait object?